src/main/java/bio/terra/service/filedata/DrsService.java

```diff
@@ -452,4 +448,41 @@ public static String getLastNameFromPath(String path) {
     String[] pathParts = StringUtils.split(path, '/');
     return pathParts[pathParts.length - 1];
   }
+
+  private void verifyAuthorization(String snapshotId, AuthenticatedUserRequest authUser) {
+    samAuthorizations.computeIfAbsent(
+        snapshotId,
+        id -> {
+          samService.verifyAuthorization(
+              authUser, IamResourceType.DATASNAPSHOT, id, IamAction.READ_DATA);
+          return id;
+        });
+  }
+
+  private SnapshotProject getSnapshotProject(UUID snapshotId) {
+    return snapshotProjects.computeIfAbsent(
+        snapshotId, snapshotService::retrieveAvailableSnapshotProject);
+  }
+
+  private SnapshotCacheResult getSnapshot(UUID snapshotId) {
+    return snapshotCache.computeIfAbsent(
+        snapshotId, id -> new SnapshotCacheResult(snapshotService.retrieve(id)));
+  }
+
+  private static class SnapshotCacheResult {
```

Comment:
I'm curious...why not just cache the snapshot?

Reply:
Snapshot is huge, and we only need a teeny tiny bit of it. The best cache elements are small cache elements!

Reply:
Fair point.  And as I think about it, it's contextual (e.g. you can ask for more or less info) and can have expiring values (e.g. signed urls to parquet files).  All that to say,g good point